### PR TITLE
kpt init command updates Kptfile inventory parameters

### PR DESCRIPTION
**Pull request by seans3**
* New `kpt live init` command set the `Inventory` parameters in the `Kptfile`.
* Not connected into `kpt` command yet.
* Adds unit tests for `KptInitOptions`.

### CODE REVIEW

File: commands/initcmd.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+// NewKptInitOptions returns a pointer to an initial KptInitOptions structure.
+func NewKptInitOptions(f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *KptInitOptions {
+	io := config.NewInitOptions(f, ioStreams)
+	return &KptInitOptions{
+		defaultInitOptions: io,
+	}
+}
+
+// Complete fills in fields for KptInitOptions based on the passed "args".
+func (io *KptInitOptions) Complete(args []string) error {
```
Comment (mortent) [inline commands/initcmd.go:43]: I'm not sure why this is split out into `Complete` and `Run` functions. Seems like we can do all in one.

File: commands/initcmd.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	force              bool   // Set inventory values even if already set in Kptfile
+	name               string // Inventory object name
+	namespace          string // Inventory object namespace
+	inventoryID        string // Inventory object unique identifier label
+	defaultInitOptions *config.InitOptions
+}
+
+// NewKptInitOptions returns a pointer to an initial KptInitOptions structure.
+func NewKptInitOptions(f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *KptInitOptions {
+	io := config.NewInitOptions(f, ioStreams)
```
Comment (mortent) [inline commands/initcmd.go:36]: The naming here is a little confusing. Can we chose a different name on this variable than we use in the function definitions. It is the same name, but different types.

File: commands/initcmd.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+// KptInitOptions encapsulates fields for kpt init command. This init command
+// fills in inventory values in the Kptfile.
+type KptInitOptions struct {
+	dir                string // Directory with Kptfile
+	force              bool   // Set inventory values even if already set in Kptfile
+	name               string // Inventory object name
+	namespace          string // Inventory object namespace
+	inventoryID        string // Inventory object unique identifier label
+	defaultInitOptions *config.InitOptions
```
Comment (mortent) [inline commands/initcmd.go:31]: I don't understand why we are nesting the old inventory template version inside the new Kptfile version. If we need to share code, we should just split it out into a separate library that can be shared.

File: commands/initcmd.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		return err
+	}
+	io.dir = io.defaultInitOptions.Dir
+	// Add random suffix to default inventory name to avoid collisions
+	if io.name == defaultInventoryName {
+		randomSuffix := common.RandomStr(time.Now().UTC().UnixNano())
+		io.name = fmt.Sprintf("%s-%s", defaultInventoryName, randomSuffix)
+	}
+	io.namespace = strings.TrimSpace(io.defaultInitOptions.Namespace)
+	io.inventoryID = strings.TrimSpace(io.defaultInitOptions.InventoryID)
```
Comment (mortent) [inline commands/initcmd.go:54]: So the id as we want to use for the Kptfile version should be based on the actual values, and not just a UID like we do with the inventory template solution.

Comment (seans3) [reply commands/initcmd.go:54]: Please explain this. I do not understand your comment.

Comment (mortent) [reply commands/initcmd.go:54]: Based on the discussions and the design, we want the generated id to be a combination of the hash of the name+namespace and the timestamp when this command is run. That will allow us to use the id as a checksum and detect situations when either the name or the namespace has been changed.

File: commands/initcmd.go (hunk 1 above)
Comment (seans3) [reply commands/initcmd.go]: Fixed.

File: commands/initcmd.go (hunk 2 above)
Comment (seans3) [reply commands/initcmd.go]: Fixed.

File: commands/initcmd.go (hunk 3 above)
Comment (seans3) [reply commands/initcmd.go]: Fixed.

File: commands/initcmd.go (hunk 4 above)
Comment (seans3) [reply commands/initcmd.go]: Initial implementation of `generateID()` function to create the `namespace/name`+`unixtimestamp` identifier (plus unit tests).

**Review by mortent — APPROVED** [review]

**Pull request merged**